Bearing displacement

ashlleysam opened the pull request:
Close #70

Commit:
Added min_fix_displacement ROS param

Commit:
Updated bearing method, changed constant name

Commit:
Fixed code formatting

irapha commented:
When this is ready for review, just put the "needs review" label to lmk and I'll look at it

irapha commented on buzzmobile/process/bearing/bearing.py:
```diff
@@ -31,7 +31,16 @@ def median(self):
+    publishes the new median.
+    """
     global last_fix
     global med_filter
```
Idk how these two lines passed code review. Can you please remove these `global` lines?

irapha commented on buzzmobile/constants.yaml:
```diff
@@ -13,7 +13,12 @@ pixels_per_m: 80
+# Minimum distance a positional fix must be from the previous
+# fix in order to add its bearing to the median filter. Used to
+# improve bearing accuracy at low speeds.
+min_fix_distance: 0.1
```
Can you add the unit of this distance to the comment?

irapha replied:
Also, depending on the units this might be too large a distance, no? If it's radians, then the min distance is enormous for a sphere the size of the earth.

irapha commented on buzzmobile/process/bearing/bearing.py:
```diff
@@ -40,13 +49,25 @@ def bearing(fix):
+        c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
+        distance = EARTH_RADIUS * c
+
+        if distance >= MIN_FIX_DISTANCE:
```
I'd say we should also save the new fix if we currently don't have any bearing.

Imagine the scenario: we turn BM on, it's stationary. Everything is waiting on the bearing to be published so they can start planning. But bearing is not being published because it's waiting on BM to start moving so it can get a good bearing.

TLDR: when we don't have information, save whatever information we can get.

irapha replied:
And there's probably a way to make the "threshhold" for what is "too low a distance" automatically calculated based on the information we already have. This won't be a part of this issue but it's something to think about. I'll open an issue for it.

irapha replied:
This last part being tracked in #78

ashlleysam replied:
So the fix is saved when there's no bearing, but the conditions for it to ~~publish a~~ add a new bearing are that there are at least 2 fixes recorded, and they're at least the delta apart. If a bearing being published is a wait condition should it just ~~publish~~ add 0 radians if it doesn't have data?

ashlleysam replied:
So assuming the delta is not met means that the car is going too slow for bearings to be accurate I would say it should always be ~~publishing a 0~~ adding a 0 to the filter, and publishing the result anyways.

ashlleysam replied:
I totally forgot that a bearing isn't relative, so I actually don't know what to do if there isn't a bearing yet

irapha replied:
I think there's this idea of a good bearing (meets the delta requirement) and a bad bearing (does not meet delta requirement)

Use last fix to update either good or bad bearing.

If there is a last good bearing,  publish that (unless its way too old, but this time part can be implemented when we're fixing #78 )

If there isnt a good bearing, publish the last bearing we were able to get (which can be a bad bearing)

Idk how exactly that translates into implementation, but we want to prioritize good bearings, but not completely fail if we don't have a good bearing. I also don't like the idea of defaulting to 0 rad bearing because our fixes, albeit noisy, might still provide a better bearing than a hardcoded 0 rad.

irapha replied:
If there isnt a bearing, dont publish, but save the fixes so we end up with a bearing at one point (even a bad one, which is better than not having any bearing)

ashlleysam replied:
Alright I understand that. I think we could change the median filter to always pop the oldest value, then when there are no more good bearings just rely on the most recent one.

irapha replied:
Idk if we want that. The purpose of the median filter is to remove noise by outputting the median bearing of the last n bearings we recorded.

Maybe I got confused before and didn't read things properly, but I think the idea behind the issue you're fixing in this PR is to save the `last_fix` (not bearing) if and only if the `last_fix` is greater than a epsilon of the `new_fix`.

Then, following this discussion, if it's _not_ greater than this epsilon _AND_ we have not filled the array in the median filter with bearing calculations, then we should drop the fix.

But we should still be publishing a bearing whenever there is a bearing in the median filter.

irapha replied:
I think my spiel about good and bad bearings was a bad spiel. The median filter is doing that partially.

irapha commented on buzzmobile/process/bearing/bearing.py:
```diff
@@ -40,13 +49,25 @@ def bearing(fix):
 
-        bearing_pub.publish(bearings['med_filter'].median())
+        #Calculates great-circle distance between positions
+        a = (math.pow(math.sin(lat2 - lat1), 2)
```
Can you add units to the comment? Is a in meters? radians? km?

Also maybe factor out this great-distance calculation.

ashlleysam replied:
Sorry, what do you mean by factor out the calculation

irapha replied:
I mean creating a new method; something like:

``` python
a = get_great_circle_distance(lat1, lat2, etc)
```

but with a better method name

irapha commented on buzzmobile/process/bearing/bearing.py:
```diff
@@ -40,13 +49,25 @@ def bearing(fix):
+            x = (math.cos(lat1) * math.sin(lat2)
+                 - math.sin(lat1) * math.cos(lat2) * math.cos(lon2 - lon1))
+            angle = math.atan2(y, x)
+            computedBearing = (angle + 2 * math.pi) % (2 * math.pi)
```
Same for this (units + factor out)?

Commit:
Units

Add units to min_fix_distance constant, add units in bearing method
calculations

Commit:
Factor out distance and angle calulations

Commit:
Fix bearing

Commit:
Fix bearing

Fixed method and updated documentation

Commit:
Method cleanup

Moved from global dictionary to variables for better comprehension,
updated method description, simplified get_distance and
get_forward_angle signatures

Commit:
Update global dict for clarity

irapha commented on buzzmobile/process/bearing/bearing.py:
```diff
@@ -25,10 +25,10 @@ def median(self):
+# Global variables
+globalVars = {}
+globalVars['last_fix'] = None
+globalVars['med_filter'] = MedianFilter(rospy.get_param('median_filter_size'))
```
nit: change to `globals`

ashlleysam replied:
OK cool. I haven't done Python before so please tell me of any stylistic or whatever errors. I understand what the dict was doing when Dip2Dip originally wrote it when he had it kinda badly named as bearings, but when you change the name to `globals` it is highlighted by the IDE? Is `globals` a reserved word or a special dict or something?

irapha commented on buzzmobile/process/bearing/bearing.py:
```diff
@@ -25,31 +25,71 @@ def median(self):
+                last_fix = fix
+        else:
+            last_fix = fix
+    globalVars['last_fix'] = last_fix
```
nit: use the dict inline. (as in, instead of doing `last_fix = globals['last']` then doing `globals['last'] = last_fix`, just use `globals['last']` everywhere

ashlleysam replied:
Also why do we leave the rsopy publisher outside of the globals dict?

irapha replied:
Because it doesn't get modified. The problem with global variables and ros is that if you have the following:

``` python
a_global

def message_cb(new_msg):
    a_global = new_msg
```

What's actually happening is that the `a_global` inside `message_cb` is not the global variable `a_global`, but a local variable to the method `message_cb`. This is not a problem with ros publishers because:

``` python
pub = rospy.Publisher('msg_name', String, queue_size=1)

def some_method():
    pub.publish('hey')
```

`pub.publish()` would fails if `pub` is a local variable, so python looks up if there's a global named `pub` and finds the publisher we defined.

I don't know why the first problem happens exactly, but it has to do with ros+python interactions because it doesn't happen in vanilla python.

joshuamorton replied:
rapha no!

Its a scoping rule. If a value is ever assigned inside a function, it is assigned a local varname and a value (see `some_method.__code__.co_varnames`). If it is never assigned to, then a recursive lookup is used to parent (nonlocal) function and finally the global scope.

That's why the result of 

```
a = 5
def func():
    a = 7
func()
print(a)
```

will be 5, and not 7, and also why

```
a = 5
def f():
    b = a
    a = 5
f()
```

will raise UnboundLocalError. Depsite the fact that `a` exists at runtime, at compile time (that is, when the function was compiled), it was created as a local. And that local isn't yet bound in the function.

To fix, use the `global` keyword.

```
a_global
def message_cb(new_msg):
    global a_global
    a_global = new_msg
```

This instructs python to use the global version of `a_global` in this function instead of creating a local.

irapha reviewed (changes requested):
two tiny nits but I looked over the actual logic and it looks good. Sorry about the confusion earlier. Your factoring out helped a lot.

irapha commented:
I dont think `globals` is a reserved word by Python. Just `global` is. So
you can use. Also when we have the pylint working it will yell at you for
stylist errors so you can learn those by the frustration of pylint :)

On Thu, Oct 27, 2016, 10:52 Roland (William) Samuelson <
notifications@github.com> wrote:

> ## _@williamsam1_ commented on this pull request.
> 
> In buzzmobile/process/bearing/bearing.py
> https://github.com/gtagency/buzzmobile/pull/77:
> 
> > @@ -25,10 +25,10 @@ def median(self):
> >              return (sortedArray[index] + sortedArray[index + 1]) / 2.0
> 
> -#Global Variables
> -bearings = {}
> -bearings['last_fix'] = None
> -bearings['med_filter'] = MedianFilter(rospy.get_param('median_filter_size'))
> +# Global variables
> +globalVars = {}
> +globalVars['last_fix'] = None
> +globalVars['med_filter'] = MedianFilter(rospy.get_param('median_filter_size'))
> 
> OK cool. I haven't done Python before so please tell me of any stylistic
> or whatever errors. I understand what the dict was doing when Dip2Dip
> originally wrote it when he had it kinda badly named as bearings, but when
> you change the name to globals it is highlighted by the IDE? Is globals a
> reserved word or a special dict or something?
> 
> —
> You are receiving this because you commented.
> 
> Reply to this email directly, view it on GitHub
> https://github.com/gtagency/buzzmobile/pull/77, or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHC0h_oOt7ynbJ_WPj63SobPXJhXD_0yks5q4LqrgaJpZM4KhPq9
> .

Commit:
Nits

Change dict to globals, dict used inline

irapha commented on buzzmobile/process/bearing/bearing.py:
```diff
@@ -43,8 +43,8 @@ def bearing(fix):
 
-    last_fix = globalVars['last_fix']
-    med_filter = globalVars['med_filter']
+    last_fix = globals['last_fix']
```
nit: you can also use `globals['last_fix']` inline in the if statement then you can remove this line.

Commit:
Nit deux

All global dicts now inline

ashlleysam commented:
Okay the Python god has spoken and not to rat out Rapha or anything but he
had me remove usages of the global keyword in that method

On Thu, Oct 27, 2016, 7:24 PM Josh Morton notifications@github.com wrote:

> ## _@joshuamorton_ commented on this pull request.
> 
> In buzzmobile/process/bearing/bearing.py
> https://github.com/gtagency/buzzmobile/pull/77:
> 
> > ```
> >  if fix is not None:
> > ```
> > -        last_fix = fix
> > -        if last_fix is not None:
> > -            distance = get_distance(last_fix, fix)
> > -            bearing = get_forward_angle(last_fix, fix)
> > -            med_filter.add(bearing)
> > -            bearing_pub.publish(med_filter.median())
> > -            if distance >= MIN_FIX_DISTANCE:
> > -                last_fix = fix
> > -        else:
> > -            last_fix = fix
> > -    globalVars['last_fix'] = last_fix
> 
> rapha no!
> 
> Its a scoping rule. If a value is ever assigned inside a function, it is
> assigned a local varname and a value (see some_method.**code**.co_varnames).
> If it is never assigned to, then a recursive lookup is used to parent
> (nonlocal) function and finally the global scope.
> 
> That's why the result of
> 
> a = 5
> def func():
>     a = 7
> func()
> print(a)
> 
> will be 5, and not 7, and also why
> 
> a = 5
> def f():
>     b = a
>     a = 5
> f()
> 
> will raise UnboundLocalError. Depsite the fact that a exists at runtime,
> at compile time (that is, when the function was compiled), it was created
> as a local. And that local isn't yet bound in the function.
> 
> To fix, use the global keyword.
> 
> a_global
> def message_cb(new_msg):
>     global a_global
>     a_global = new_msg
> 
> This instructs python to use the global version of a_global in this
> function instead of creating a local.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/gtagency/buzzmobile/pull/77, or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHpP2sfnV3nkD0DSin6plMPgJplzqqwZks5q4TLGgaJpZM4KhPq9
> .

joshuamorton commented:
Well, its possible that there's a design issue. Consider using ros as a way to 'store' globals. Think, for example, having a node that subscribes to here, that keeps the last pose, and this node also subscribes to that node. IDK if you can even do that though.

irapha commented:
Wait if I'm being ratted out I'll point out that I had consulted Josh about this and we both said the use of `global` was kind of icky. That's why we've been doing dicts, because that works.

I believe Josh's "rapha no" comment was bc my explanation was wrong, correct?

We can look up if ros has a better way of storing globals but I'm almost 100% sure the answer is no. Until then I'll say dicts are still the way we're using to store globals, because it works.

irapha commented:
~~This being said, if we want to move towards using `global a_global`, I'd say open an issue and do it in a separate PR because it will modify way more than `bearing.py`~~

Not doing this. See below

joshuamorton commented:
Correct, explanation was wrong. Globals are stil icky.

On Thu, Oct 27, 2016 at 8:37 PM Raphael Gontijo Lopes <
notifications@github.com> wrote:

> Wait if I'm being ratted out I'll point out that I had consulted Josh
> about this and we both said the use of global was kind of icky. That's
> why we've been doing dicts, because that works.
> 
> I believe Josh's "rapha no" comment was bc my explanation was wrong,
> correct?
> 
> We can look up if ros has a better way of storing globals but I'm almost
> 100% sure the answer is no. Until then I'll say dicts are still the way
> we're using to store globals, because it works.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/gtagency/buzzmobile/pull/77#issuecomment-256807777,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AEz4AH7TD-NiKEa1wJYjXicYbxibTCerks5q4UOsgaJpZM4KhPq9
> .

ashlleysam commented:
Why did my one method low hanging fruit pr have to spawn a whole
application structure discussion -_-

On Fri, Oct 28, 2016 at 8:39 AM Josh Morton notifications@github.com
wrote:

> Correct, explanation was wrong. Globals are stil icky.
> 
> On Thu, Oct 27, 2016 at 8:37 PM Raphael Gontijo Lopes <
> notifications@github.com> wrote:
> 
> > Wait if I'm being ratted out I'll point out that I had consulted Josh
> > about this and we both said the use of global was kind of icky. That's
> > why we've been doing dicts, because that works.
> > 
> > I believe Josh's "rapha no" comment was bc my explanation was wrong,
> > correct?
> > 
> > We can look up if ros has a better way of storing globals but I'm almost
> > 100% sure the answer is no. Until then I'll say dicts are still the way
> > we're using to store globals, because it works.
> > 
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > https://github.com/gtagency/buzzmobile/pull/77#issuecomment-256807777,
> > or mute the thread
> > <
> > https://github.com/notifications/unsubscribe-auth/AEz4AH7TD-NiKEa1wJYjXicYbxibTCerks5q4UOsgaJpZM4KhPq9
> > 
> > .
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/gtagency/buzzmobile/pull/77#issuecomment-256808177,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AHpP2trEyQOrdRBkZE9zuVqz0rEjBwZZks5q4URcgaJpZM4KhPq9
> .

irapha commented on buzzmobile/process/bearing/bearing.py:
```diff
@@ -25,31 +25,67 @@ def median(self):
-bearings['last_fix'] = None
-bearings['med_filter'] = MedianFilter(rospy.get_param('median_filter_size'))
+# Globals
+globals = {}
```
Change to `g = {} # globals`

Commit:
Change name of global dict

Commit:
Merge remote-tracking branch 'refs/remotes/origin/master' into bearing
